### PR TITLE
NAS-119265 / 23.10 / New `ix-table2` component

### PR DESCRIPTION
**Pull request by bvasilenko**
**Summary:**

## Usages
This PR includes usages of a new table component:
I. at **System Settings > Advanced > Sessions**  (simple)
II. at **Credentials > Local Users** (advanced)

## I. Simple usage of  `ix-table2`:

1. `IxTable2Component<T>` new component, not connected to `ix-table` or `MatTable`. `ix-table2` renders `table` with all contents inside it

```
<ix-table2>

  <!-- use static headers. we'll replace them in "II" -->
  <tr>
    <th>First name</th>
    <th>Last name</th>
  </tr>

  <ix-table-body [columns]="columns" [dataProvider]="dataProvider"></ix-table-body>
</ix-table2>
```

2. `ix-table-body` renders rows and cells. `@Input() columns` accepts an array, for example:

```
columns = [
  'firstName', // type safe
  'lastName',
  {
     title: 'Age',
     propertyName: 'dateOfBirth', // type safe
     sortBy: (row) => getAge(row.dateOfBirth),
  }
]
```

3. No `rows` array, but `@Input() dataProvider` accepts an object of `DataProvider<T>` interface

```
DataProvider<T> {
   currentPage: Observable<T[]>; // emits array of rows
}
```

4. basic reusable provider (included into PR)  `class ArrayDataProvider extends DataProvider { ... }`

5.  `ArrayDataProvider`  accepts a static array:

```
const rows = [
...
]
dataProvider = new ArrayDataProvider(rows);
```

6.  **cell template** uses `ng-template`, to render custom cell contents. Example of usage: "actions" column.

## II. Advanced `ix-table2` usages:

**Headers & sorting:**

1. with `ix-table-head` component or directive, you can render heading cells

```
<ix-table2>
  <ix-table-head [columns]="columns" [dataProvider]="dataProvider"></ix-table-head>

  <ix-table-body [columns]="columns" [dataProvider]="dataProvider"></ix-table-body>
</ix-table2>
```

2. when user clicks a column header, `currentPage` is sorted with  using `_.sortBy` if supplied, or with `propertName` otherwise

```
columns = [
  'firstName', // no `sortBy`, sorting is applied by the prop (`firstName`)
  'lastName', // no `sortBy`, same as above
  {
     sortBy: (row) => getAge(row.dateOfBirth),
  }
]
```

3. Sorting algorithm is encapsulated by `dataProvider` :

```
DataProvider<T> {
   setSorting({
       string, // type safe
       desc: boolean,
   });
   currentPage: Observable<T[]>; // re-emits on 'setSorting'
   sorting: { ... } // read-only
}
```

**Pagination:**

1. use `ix-table-pager` to render "page size" dropdown and `<` `>` buttons

```
<ix-table2>
  ...
  <ix-table-pager [dataProvider]="dataProvider"></ix-table-pager>
</ix-table2>
```

2. pagination logic is encapsulated by `dataProvider`

```
DataProvider<T> {
   ...
   setPagination({
       pageNumber,
       pageSize,
   });
   currentPage: Observable<T[]>; // array of rows, re-emit on 'setPagination'
   ...
   pagination: { ... } // read-only
}
```

**Row details block:**

Pass another `ng-template` to `ix-table-body`. For `row.isExpanded`, additional fully spanned row is rendered with the specified `ng-template`.

### CODE REVIEW

File: src/app/pages/system/advanced/sessions/sessions-card/sessions-card.component.ts
Comment (undsoft) [inline src/app/pages/system/advanced/sessions/sessions-card/sessions-card.component.ts:79]: I'm not a huge fan of doing transformations like this. You need a new interface and it tends to get messy. I'd rather we keep original interface and format data in table-related code.

File: src/app/modules/ix-table2/array-data-provider.ts
Comment (undsoft) [inline src/app/modules/ix-table2/array-data-provider.ts:24]: I feel like ArrayDataProvider shouldn't be accepting data as observables and should just accept static values.
In the future we may consider having something like `AsyncDataProvider` that would accept `this.ws.call()` directly.

File: src/app/pages/account/users/user-list/user-list.component.html
Comment (undsoft) [inline src/app/pages/account/users/user-list/user-list.component.html:37]: In the future it would be good to think about how to make this more reusable, for example, by allowing column definitions to accept classes and having `BooleanColumn` class.

File: src/app/pages/account/users/user-list/user-list.component.spec.ts
Comment (undsoft) [inline src/app/pages/account/users/user-list/user-list.component.spec.ts:126]: As a separate task we should create harnesses to use in tests.

File: src/app/modules/ix-table2/components/ix-table-head/ix-table-head.component.html
Comment (undsoft) [inline src/app/modules/ix-table2/components/ix-table-head/ix-table-head.component.html:7]: Sounds like a time for `SortDirection` enum.

**Review by undsoft — APPROVED** [review]
Looks good overall. Syntax appears easier to use than what we have now.
Let's wait for more approval before proceeding.

Issues to fix:
1. User table does not support filtering and likely doesn't handle empty states. Empty states may or may not be part of the table components, not sure.
2. Would be good to see examples with how to set default sorting on user table.

File: src/app/modules/ix-table2/components/ix-table-head/ix-table-head.component.html
Comment (denysbutenko) [inline src/app/modules/ix-table2/components/ix-table-head/ix-table-head.component.html:14]: Is there going to be something? It seems like ngIf would be enough.

**Review by denysbutenko — APPROVED** [review]
Looks good to me.

**Review by RehanY147 — APPROVED** [review]

File: src/app/modules/ix-table2/components/ix-table-pager/ix-table-pager.component.html
Comment (dszidi) [inline src/app/modules/ix-table2/components/ix-table-pager/ix-table-pager.component.html:1]: Seems there is a way to set the `itemsPerPage` in the `ix-table-pager.component.ts` file but nothing in the pager's UI itself. This setting should really be exposed to the user in the pager's UI.

**Review by dszidi — CHANGES_REQUESTED** [review]

File: src/app/pages/account/users/user-list/user-list.component.spec.ts
Comment (bvasilenko) [reply src/app/pages/account/users/user-list/user-list.component.spec.ts:126]: https://ixsystems.atlassian.net/browse/NAS-122527

File: src/app/pages/account/users/user-list/user-list.component.html
Comment (bvasilenko) [reply src/app/pages/account/users/user-list/user-list.component.html:37]: https://ixsystems.atlassian.net/browse/NAS-122529

**Review by AlexKarpov98 — APPROVED** [review]
Nice 👍

**Review by dszidi — APPROVED** [review]
LGTM

**Comment by undsoft** [comment]
5 approvals out of 5

**Pull request merged**
